pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *ControllerV2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *ControllerV2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.AddAfter(key, t)
+}
+
+// updateCronJob re-queues the CronJob for next scheduled time if there is a
+// change in spec.schedule otherwise it re-queues it now
+func (jm *ControllerV2) updateCronJob(old interface{}, curr interface{}) {
+	oldCJ, okOld := old.(*batchv1beta1.CronJob)
+	newCJ, okNew := curr.(*batchv1beta1.CronJob)
+
+	if !okOld || !okNew {
+		// typecasting of one failed, handle this better, may be log entry
+		return
+	}
+	// if the change in schedule results in next requeue having to be sooner than it already was,
+	// it will be handled here by the queue. If the next requeue is further than previous schedule,
+	// the sync loop will essentially be a no-op for the already queued key with old schedule.
+	if oldCJ.Spec.Schedule != newCJ.Spec.Schedule {
+		// schedule changed, change the requeue time
+		sched, err := cron.ParseStandard(newCJ.Spec.Schedule)
+		if err != nil {
+			// this is likely a user error in defining the spec value
+			// we should log the error and not reconcile this cronjob until an update to spec
+			klog.V(2).InfoS("unparseable schedule for cronjob", "cronjob", klog.KRef(newCJ.GetNamespace(), newCJ.GetName()), "schedule", newCJ.Spec.Schedule, "err", err)
+			jm.recorder.Eventf(newCJ, corev1.EventTypeWarning, "UnParseableCronJobSchedule", "unparseable schedule for cronjob: %s", newCJ.Spec.Schedule)
+			return
+		}
+		now := jm.now()
+		t := nextScheduledTimeDuration(sched, now)
+
+		jm.enqueueControllerWithTime(curr, *t)
+		return
+	}
+
+	// other parameters changed, requeue this now and if this gets triggered
+	// within deadline, sync loop will work on the CJ otherwise updates will be handled
+	// during the next schedule
+	// TODO: need to handle the change of spec.JobTemplate.metadata.labels explicitly
+	//   to cleanup jobs with old labels
+	jm.enqueueController(curr)
+}
+
+// TODO: @alpatel we need to return errors from this function, in order to allow
+//  for errors to propagate up the reconcile function and kick in queues exponential
+//  backed off retries more details on inline code comments in the func
+
+// syncOne reconciles a CronJob with a list of any Jobs that it created.
+// All known jobs created by "cj" should be included in "js".
+// The current time is passed in to facilitate testing.
+// It has no receiver, to facilitate testing.
+func syncOne2(
+	cj *batchv1beta1.CronJob,
+	js []batchv1.Job,
+	now time.Time,
```

Comment:
I would have expected a fakeclock on the `jm`

Reply:
+1, `jm` already had a function, updated to use it. I realize I can make it an interface if need be, lmk. I figured I would keep it as it since it's only needed for testing.